Sqrt price lower upper

jeffishjeff opened the pull request.

jeffishjeff committed:
use sqrtPrice (uint160) for tick key instead of uint32

jeffishjeff committed:
simplify order id by making it a full word

jeffishjeff committed:
unit test for PoolKey

wangshouh committed:
Finish OrderId library and test

wangshouh committed:
Modify pools to variables

wangshouh committed:
Delete _getPools

jeffishjeff committed:
minor change to get the pr merged

jeffishjeff committed:
#9

jeffishjeff commented:
issue #9

jeffishjeff committed:
Position custom type

jeffishjeff committed:
Merge branch 'main' into sqrtPrice-lower-upper

wangshouh commented on src/models/Position.sol:
```diff
@@ -0,0 +1,61 @@
+                    )
+                )
+        }
+    }
```
Split using the `set` function, like `setLiquidity`/ `setRangeRatio` / `setThresholdRatio`

jeffishjeff committed:
revamp

jeffishjeff commented:
ended up doing a full revamp of code to date:

- a new SqrtPrice type that'll encapsulate x96 logics later
- pool references an external liquidity token
- modifyLiquidity -> v2 style mint/burn, though still using flash accounting pattern
- pool tracks reserve0/1, liquidity renamed to market depth; these are necessary as market depth changes during rebalancing
- added events to IPoolManager
- renamed tick to sqrt price level, to be more consistent w/ trad-fi lingo

ready for review @wangshouh

wangshouh commented on src/interfaces/IPoolManager.sol:
```diff
@@ -1,63 +1,110 @@
+        returns (BalanceDelta balanceDelta, int128 liquidityDelta);
 
-    // TODO: do we need additional params to specify the sqrtPrice and tick where order will created or whatever sqrtPriceLimit is fine?
+    // TODO: add price limit, to decide whether use uint32 or SqrtPrice
+    // TODO: add neighbor ticks/SqrtPrices that are already in the linked map, to decide whether use uint32 or SqrtPrice
```
In the previous discussion, price limit uses `uint32`, and neighbor SqrtPirces uses `SqrtPrice`. Has this conclusion changed?

jeffishjeff replied:
Yeah, we talked about that but I thought it wasn't fully decided, depending on how we implement tick (a.k.a SqrtPriceLevel) rules and math lib implementation.

For example, if we go w/ the algo that rounds down to 1 / 2^20 of SqrtPrice (i.e. drop everything to the right of 20th sig fig). Then we don't need uint32 or require price limit to be "on the tick".

But this is your part so I yield to you :)

wangshouh commented on src/PoolManager.sol:
```diff
@@ -7,34 +7,46 @@ import {OrderId} from "./models/OrderId.sol";
     /// @inheritdoc IPoolManager
     function placeOrder(PoolKey calldata poolKey, PlaceOrderParams calldata params)
         external
-        returns (BalanceDelta balanceDelta, OrderId orderId)
+        returns (OrderId orderId, BalanceDelta balanceDelta)
     {
-        // TODO: implement
+        // TODO: Implement
     }
```
If placeOrder is fully filled and don't need place order in LOB, how should we return OrderId? Should we directly return an OrderId consisting of 0?

jeffishjeff replied:
Yeah I think so

wangshouh commented on src/models/SqrtPriceLevel.sol:
```diff
@@ -3,12 +3,13 @@ pragma solidity ^0.8.0;
+    SqrtPrice prev;
+    SqrtPrice next;
+    uint128 totalOpenAmount;
+    uint64 lastOpenOder;
```
Type: `lastOpenOrder`

jeffishjeff replied:
fixed

jeffishjeff committed:
typo
